packages/yew-functional/src/hooks/use_reducer.rs

```diff
@@ -99,47 +104,75 @@ where
     reducer: Reducer,
     initial_state: InitialState,
     init: InitFn,
-) -> (Rc<State>, Rc<impl Fn(Action)>)
+) -> UseReducerHandle<State, Action>
 where
     Reducer: Fn(Rc<State>, Action) -> State + 'static,
-    InitFn: Fn(InitialState) -> State,
+    Action: 'static,
+    State: 'static,
+    InitialState: 'static,
+    InitFn: Fn(InitialState) -> State + 'static,
 {
-    struct UseReducerState<State> {
-        current_state: Rc<State>,
-    }
-    impl<T> Hook for UseReducerState<T> {}
     let init = Box::new(init);
     let reducer = Rc::new(reducer);
     use_hook(
-        |internal_hook_change: &mut UseReducerState<State>, hook_callback| {
-            (
-                internal_hook_change.current_state.clone(),
-                Rc::new(move |action: Action| {
-                    let reducer = reducer.clone();
-                    hook_callback(
-                        move |internal_hook_change: &mut UseReducerState<State>| {
-                            internal_hook_change.current_state = Rc::new((reducer)(
-                                internal_hook_change.current_state.clone(),
-                                action,
-                            ));
-                            true
-                        },
-                        false, // run pre render
-                    );
-                }),
-            )
-        },
-        move || UseReducerState {
+        move || UseReducer {
             current_state: Rc::new(init(initial_state)),
         },
+        |s, updater| {
+            let setter: Rc<dyn Fn(Action)> = Rc::new(move |action: Action| {
+                let reducer = reducer.clone();
+                // We call the callback, consumer the updater
+                // Required to put the type annotations on Self so the method knows how to downcast
+                updater.callback(move |state: &mut UseReducer<State>| {
+                    let new_state = reducer(state.current_state.clone(), action);
+                    state.current_state = Rc::new(new_state);
+                    true
+                });
+            });
+
+            UseReducerHandle {
+                value: Rc::clone(&s.current_state),
+                setter,
+            }
+        },
+        |_| {},
     )
 }
+
+/// State handle for [`use_reducer`] hook
+pub struct UseReducerHandle<STATE, ACTION> {
+    value: Rc<STATE>,
+    setter: Rc<dyn Fn(ACTION)>,
+}
+
+impl<STATE, ACTION> UseReducerHandle<STATE, ACTION> {
+    pub fn dispatch(&self, value: ACTION) {
+        (self.setter)(value)
+    }
+}
+
+impl<STATE, ACTION> Deref for UseReducerHandle<STATE, ACTION> {
+    type Target = STATE;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.value
+    }
+}
+
+impl<STATE, ACTION> Clone for UseReducerHandle<STATE, ACTION> {
+    fn clone(&self) -> Self {
+        Self {
+            value: Rc::clone(&self.value),
+            setter: Rc::clone(&self.setter),
+        }
+    }
+}
```

Comment:
Per <https://rust-lang.github.io/api-guidelines/naming.html>, type parameters should use UpperCamelCase

```suggestion
pub struct UseReducerHandle<State, Action> {
    value: Rc<State>,
    setter: Rc<dyn Fn(Action)>,
}

impl<State, Action> UseReducerHandle<State, Action> {
    pub fn dispatch(&self, value: Action) {
        (self.setter)(value)
    }
}

impl<State, Action> Deref for UseReducerHandle<State, Action> {
    type Target = State;

    fn deref(&self) -> &Self::Target {
        &*self.value
    }
}

impl<State, Action> Clone for UseReducerHandle<State, Action> {
    fn clone(&self) -> Self {
        Self {
            value: Rc::clone(&self.value),
            setter: Rc::clone(&self.setter),
        }
    }
}
```